File: GenMotionEasy/Motion/MotionControlManager.cs
using GenMotionEasy.Motion.Control;
using GTN;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace GenMotionEasy.Motion
{
    public class MotionControlManager
    {
        private readonly Dictionary<int, AxisController> _axisControllers = new Dictionary<int, AxisController>();
        private short _core = 1;

        private readonly object _lock = new object();

        public MotionControlManager()
        {

        }


        public bool Open()
        {
            if (mc.GTN_Open(5, 1) != 0)
                return false;
            mc.GTN_Reset(_core);
            return true;
        }

        public short EcatLoad()
        {
            mc.GTN_TerminateEcatComm(_core);
            return mc.GTN_InitEcatComm(_core);
        }

        public short EcatState(out short state)
        {
            return mc.GTN_IsEcatReady(_core, out state);
        }

        public short EcatStart()
        {
            return mc.GTN_StartEcatComm(_core);
        }

        public void Close()
        {
            mc.GTN_Stop(_core, 0xFFF, 0xFFF);
        }

        public void AddAxis(short axisId, short core)
        {
            lock (_lock)
            {
                if (!_axisControllers.ContainsKey(axisId))
                {
                    _axisControllers[axisId] = new AxisController(core, axisId);
                }
            }
        }


        public AxisController GetAxisController(int axisId)
        {

            lock (_lock)
            {
                if (!_axisControllers.ContainsKey(axisId))
                {
                    throw new Exception($"轴线{axisId}不存在");
                }
                return _axisControllers[axisId];
            }
        }


        public int GetAxisCount()
        {
            return _axisControllers.Count;
        }
    }
}

File: GenMotionEasyCpp/src/MotionControlManager.cpp
#include "GenMotionEasyCpp/MotionControlManager.h"
#include <stdexcept>

MotionControlManager::MotionControlManager()
    : _lock(new int()) /* dummy lock placeholder */
    , _ecatIO(new EcatIO(_core, _lock))
    , _expansionIO(new ExpansionIO(_core, _lock))
{
}

MotionControlManager::~MotionControlManager() {
    Close();
}

bool MotionControlManager::Open() {
    if (GTN_Open(5, 2) != 0)
        return false;
    GTN_Reset(_core);
    return true;
}

short MotionControlManager::EcatLoad() {
    GTN_TerminateEcatComm(_core);
    return GTN_InitEcatComm(_core);
}

short MotionControlManager::EcatState(short& state) {
    return GTN_IsEcatReady(_core, &state);
}

short MotionControlManager::EcatStart() {
    GT_GLinkInitEx(0, 0);
    return GTN_StartEcatComm(_core);
}

void MotionControlManager::Close() {
    GTN_Stop(_core, 0xFFF, 0xFFF);
}

void MotionControlManager::AddAxis(short axisId, short core) {
    if (_axisControllers.find(axisId) == _axisControllers.end()) {
        _axisControllers[axisId] = std::make_unique<AxisController>(core, axisId);
    }
}

AxisController* MotionControlManager::GetAxisController(int axisId) {
    auto it = _axisControllers.find(axisId);
    if (it == _axisControllers.end()) {
        throw std::runtime_error("\u8f74\u7ebf" + std::to_string(axisId) + "\u4e0d\u5b58\u5728");
    }
    return it->second.get();
}

int MotionControlManager::GetAxisCount() {
    return (int)_axisControllers.size();
}
